stablehlo/transforms/StablehloRefineShapes.cpp

```diff
@@ -0,0 +1,937 @@
+
+    return evalBinary(rewriter, op, [&](APInt lhsInt, APInt rhsInt) {
+      auto result = lhsInt != 0 && rhsInt != 0;
+      return APInt(/*numBits=*/1, result);
+    });
+  }
+};
+
+struct EvalBroadcastInDimOpPattern : public OpRewritePattern<BroadcastInDimOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(BroadcastInDimOp op,
+                                PatternRewriter& rewriter) const override {
+    auto operandType = op.getOperand().getType().dyn_cast<RankedTensorType>();
+    if (!operandType || operandType.getRank() != 0)
+      return rewriter.notifyMatchFailure(op, "expected 0-dimensional type");
+
+    SmallVector<APInt> operand;
+    if (failed(matchInts(op.getOperand(), operand)))
+      return rewriter.notifyMatchFailure(op, "expected constant operands");
+    auto scalar = operand[0];
+
+    rewriter.replaceOpWithNewOp<ConstantOp>(
+        op, DenseIntElementsAttr::get(op.getResult().getType(), scalar));
+    return success();
+  }
+};
+
+struct EvalCompareOpPattern : public OpRewritePattern<CompareOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(CompareOp op,
+                                PatternRewriter& rewriter) const override {
+    auto resultType = op.getResult().getType().cast<ShapedType>();
+    auto isResultUnsigned = resultType.getElementType().isUnsignedInteger();
+    return evalBinary(rewriter, op, [&](APInt lhs, APInt rhs) {
+      bool result;
+      switch (op.getComparisonDirection()) {
+        case ComparisonDirection::EQ:
+          result = lhs == rhs;
+          break;
+        case ComparisonDirection::NE:
+          result = lhs != rhs;
+          break;
+        case ComparisonDirection::GE:
+          result = isResultUnsigned ? lhs.uge(rhs) : lhs.sge(rhs);
+          break;
+        case ComparisonDirection::GT:
+          result = isResultUnsigned ? lhs.ugt(rhs) : lhs.sgt(rhs);
+          break;
+        case ComparisonDirection::LE:
+          result = isResultUnsigned ? lhs.ule(rhs) : lhs.sle(rhs);
+          break;
+        case ComparisonDirection::LT:
+          result = isResultUnsigned ? lhs.ult(rhs) : lhs.slt(rhs);
+          break;
+        default:
+          llvm::report_fatal_error("unsupported direction");
+      }
+      return APInt(/*numBits=*/1, result);
+    });
+  }
+};
+
+struct EvalConcatenateOpPattern : public OpRewritePattern<ConcatenateOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(ConcatenateOp op,
+                                PatternRewriter& rewriter) const override {
+    auto resultType = op.getResult().getType().dyn_cast<RankedTensorType>();
+    if (!resultType || resultType.getRank() != 1)
+      return rewriter.notifyMatchFailure(op, "expected 1-dimensional type");
+
+    SmallVector<APInt> result;
+    for (Value operand : op->getOperands()) {
+      if (failed(matchInts(operand, result)))
```

Comment:
The name and behavior matches the name and behavior of upstream `matchPattern` which fills in the provided arguments, e.g. `DenseIntElementsAttr`. Given the precedent, I think it might be acceptable to leave things as they are. What do you think?

Reply:
I see. That reminds me of the arg `inferReturnTypes` which is also appended inside the shape function. Looks like this is an MLIR coding pattern. Let's keep it as is then!